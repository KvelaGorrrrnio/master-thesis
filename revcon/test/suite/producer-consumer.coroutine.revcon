@coroutine

coroutine main(int sum) {
  # Setup
  local int initial_memory_address 0
  spawn producer remote_memory
  spawn consumer local_sum
  spawn s stream (initial_memory_address)

  # Do stuff
  resume s (producer, consumer)

  # Get result
  resume consumer(sum)

  # Teardown
  despawn s stream (initial_memory_address)
  despawn producer remote_memory
  despawn consumer local_sum
  delocal int initial_memory_address 0
}

coroutine stream(coroutine(int,int) producer, coroutine(int) consumer) with (int initial_memory_address) {
  local int memory_address initial_memory_address
  local int data 0
  from memory_address == initial_memory_address
  do
    resume producer (memory_address, data)
    resume consumer (data)
  until memory_address == initial_memory_address
  delocal int data 0
  delocal int memory_address initial_memory_address
}

coroutine remote_memory(int pointer, int data) {
  # Setup
  local list int memory [1,2,3,4,5,6,7,8,9,10]
  local int initial_pointer pointer

  # Produce
  from initial_pointer == pointer
  loop
    data += memory[pointer]
    pointer += 1
    yield produce
    pointer -= 1
    data -= memory[pointer]
    pointer += 1
  until pointer == size memory

  # Teardown
  pointer -= size memory - initial_pointer
  delocal int initial_pointer pointer
  delocal list int memory [1,2,3,4,5,6,7,8,9,10]
}

coroutine local_sum(int data) {
  local int sum 0
  from sum == 0
  do
    sum += data
    yield consume
  until data == 0
  yield get_sum
  swap sum data
  delocal int sum 0
}
