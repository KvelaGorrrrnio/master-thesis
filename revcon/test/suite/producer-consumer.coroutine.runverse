test/suite/producer-consumer.coroutine.revcon:
[store]
  main: Ã˜
  sum: 0

[stack]
[spawn main main end () | unresume main (sum) | despawn main main begin ()] <|> []

