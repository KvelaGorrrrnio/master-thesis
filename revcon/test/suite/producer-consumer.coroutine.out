test/suite/producer-consumer.coroutine.revcon:
[store]
  main: Ã˜
  sum: 55

[stack]
[spawn main main begin () | resume main (sum) | despawn main main end ()] <|> []

